test/server/echo.go

```diff
@@ -120,3 +143,61 @@ func main() {
 	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
 	<-sigs
 }
+
+func setResponseFromCodes(request *http.Request, response http.ResponseWriter) error {
+	responseCodes := request.FormValue("codes");
+
+	codes, err := validateCodes(responseCodes)
+	if err != nil {
+		return err
+	}
+
+	// Keep a cursor for each sequence of response codes
+	position, ok := state[responseCodes]
+	if !ok {
+		state[responseCodes] = 0
+	}
+	if len(codes) > 0 {
+		state[responseCodes] = (position + 1) % len(codes)
+	}
+
+	response.WriteHeader(codes[position])
+	return nil
+}
+
+
+// codes must be comma-separated HTTP response codes
+func validateCodes(codestrings string) ([]int, error) {
+
+	if codestrings == "" {
+		// Consider no codes to be a repeating sequence of 200,
+		codestrings = strconv.Itoa(http.StatusOK)
+	}
+
+	aCodestrings := strings.Split(codestrings, ",")
+	codes := make([]int, len(aCodestrings), len(aCodestrings))
+
+	for i, codestring := range aCodestrings {
+		code, err := validateCode(codestring)
```

Comment:
I found using a map gave me a lot of flexibility.

When writing test cases against a single instance of a server I could be assured if I started a new sequence of codes I would be starting at 0.

On a previous project I spent a lot of time verifying fencepost errors.  If my server generates 17 errors, and I retry 17 times, am I guaranteed to fail?  It is easy to make that happen if each test uses a unique sequence like "501,500,500,500,200" and "502,500,500,500,500,200".  Is it worth giving that up to use an int instead of a map?